core/src/main/java/hudson/Util.java

```diff
@@ -304,19 +337,144 @@ private static void makeWritable(@Nonnull File f) {
 
     }
 
+    /**
+     * Deletes the given directory (including its contents) recursively.
+     * It does not take no for an answer - if necessary, it will have multiple
+     * attempts at deleting things.
+     *
+     * @throws IOException
+     * if the operation fails.
+     */
     public static void deleteRecursive(@Nonnull File dir) throws IOException {
+        for( int numberOfAttempts=1 ; ; numberOfAttempts++ ) {
+            try {
+                tryOnceDeleteRecursive(dir);
+                break; // success
+            } catch (IOException ex) {
+                boolean threadWasInterrupted = pauseBetweenDeletes(numberOfAttempts);
+                if( numberOfAttempts>= DELETION_MAX || threadWasInterrupted)
+                    throw new IOException(deleteFailExceptionMessage(dir, numberOfAttempts, threadWasInterrupted), ex);
+            }
```

Comment:
You are correct that the code is similar, and that it would be possible to refactor this - this was a strategy that considered at the time but I found that, instead of reducing the amount of code, the additional complexity added to the amount of code required and (more importantly) reduced the clarity of the code.
I'm of the opinion that "simple" is better than "clever", especially when there are only 3 (co-located) copies.